File: src/main/java/engineer/carrot/warren/thump/connection/ReconnectPolicy.java
package engineer.carrot.warren.thump.connection;

import engineer.carrot.warren.thump.config.ServerConfiguration;

public class ReconnectPolicy {
    private boolean shouldReconnect;
    private int reconnectDelaySeconds;
    private int maxConsecutiveReconnects;

    private boolean isDisabled = false;

    private int currentConnectionAttempt = 1;

    private final Object policyLock;

    public ReconnectPolicy(ServerConfiguration configuration) {
        this.shouldReconnect = configuration.shouldReconnectAutomatically;
        this.reconnectDelaySeconds = configuration.automaticReconnectDelaySeconds;
        this.maxConsecutiveReconnects = configuration.maxConsecutiveReconnectAttempts;

        this.policyLock = new Object();
    }

    public boolean shouldReconnect() {
        synchronized (this.policyLock) {
            return this.shouldReconnect;
        }
    }

    public int getCurrentConnectionAttempt() {
        synchronized (this.policyLock) {
            return this.currentConnectionAttempt;
        }
    }

    public void incrementConnectionAttempt() {
        synchronized (this.policyLock) {
            this.currentConnectionAttempt++;
        }
    }

    public void resetCurrentConnectionAttempts() {
        synchronized (this.policyLock) {
            this.currentConnectionAttempt = 1;
            this.isDisabled = false;
        }
    }

    public int getReconnectDelaySeconds() {
        synchronized (this.policyLock) {
            return this.reconnectDelaySeconds;
        }
    }

    public int getMaxConsecutiveReconnects() {
        synchronized (this.policyLock) {
            return this.maxConsecutiveReconnects;
        }
    }

    public boolean getIsDisabled() {
        synchronized (this.policyLock) {
            return this.isDisabled;
        }
    }

    public void disableNextReconnect() {
        synchronized (this.policyLock) {
            this.isDisabled = true;
        }
    }
}

File: src/main/kotlin/engineer/carrot/warren/thump/connection/ReconnectPolicy.kt
package engineer.carrot.warren.thump.connection

import engineer.carrot.warren.thump.config.ServerConfiguration

class ReconnectPolicy(configuration: ServerConfiguration) {
    private val shouldReconnect: Boolean
    private val reconnectDelaySeconds: Int
    private val maxConsecutiveReconnects: Int

    private var isDisabled = false

    private var currentConnectionAttempt = 1

    private val policyLock: Any

    init {
        this.shouldReconnect = configuration.shouldReconnectAutomatically
        this.reconnectDelaySeconds = configuration.automaticReconnectDelaySeconds
        this.maxConsecutiveReconnects = configuration.maxConsecutiveReconnectAttempts

        this.policyLock = Object()
    }

    fun shouldReconnect(): Boolean {
        synchronized (this.policyLock) {
            return this.shouldReconnect
        }
    }

    fun getCurrentConnectionAttempt(): Int {
        synchronized (this.policyLock) {
            return this.currentConnectionAttempt
        }
    }

    fun incrementConnectionAttempt() {
        synchronized (this.policyLock) {
            this.currentConnectionAttempt++
        }
    }

    fun resetCurrentConnectionAttempts() {
        synchronized (this.policyLock) {
            this.currentConnectionAttempt = 1
            this.isDisabled = false
        }
    }

    fun getReconnectDelaySeconds(): Int {
        synchronized (this.policyLock) {
            return this.reconnectDelaySeconds
        }
    }

    fun getMaxConsecutiveReconnects(): Int {
        synchronized (this.policyLock) {
            return this.maxConsecutiveReconnects
        }
    }

    fun getIsDisabled(): Boolean {
        synchronized (this.policyLock) {
            return this.isDisabled
        }
    }

    fun disableNextReconnect() {
        synchronized (this.policyLock) {
            this.isDisabled = true
        }
    }
}
